src/game_api/rpc.cpp

```diff
@@ -577,14 +573,25 @@ void pick_up(uint32_t who_uid, uint32_t what_uid)
     }
 }
 
-void drop(uint32_t who_uid, uint32_t what_uid)
+void drop(uint32_t who_uid, std::optional<uint32_t> what_uid)
 {
-    Movable* ent = (Movable*)get_entity_ptr(who_uid);
-    Movable* item = (Movable*)get_entity_ptr(what_uid);
-    if (ent != nullptr && item != nullptr)
+    auto ent = get_entity_ptr(who_uid);
+    if (ent == nullptr)
+        return;
+
+    if (!ent->is_movable()) // game would probably use the is_player_or_monster function here, since they are the only ones who should be able to hold something
+        return;
+
+    auto mov = ent->as<Movable>();
+    if (what_uid.has_value()) // should we handle what_uid = -1 the same way?
     {
-        ent->drop(item);
+        auto item = get_entity_ptr(what_uid.value());
+        if (item == nullptr)
+            return;
+        if (item->overlay != mov && mov->holding_uid == what_uid)
+            return;
     }
+    mov->drop();

```

Comment:
Asked this on discord but didn't get replay
this changes the behavior slightly, but makes the most sense with the parameters given

Reply:
So, did we think this virtual could be used to remove any entity from the items, but it actually just drops whatever is held when the uid is valid? Or did it have a different implementation that actually worked like it says, and was then simplified to a more broken one? Seems like an ok compromise anyway.